File: Codes in Python3/Graph/Surrounded Regions Replace 0s with Xs.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        ROWS,COLS=len(board),len(board[0])
        visited=[[False]*COLS for _ in range(ROWS)]
        direction=[[-1,0],[0,-1],[1,0],[0,1]]
        def dfs(r,c):
            if visited[r][c]:
                return
            visited[r][c]=True
            for dr,dc in direction:
                nrow,ncol=dr+r,dc+c
                if nrow in range(0,ROWS) and ncol in range(0,COLS) \
                and board[nrow][ncol]=="O" \
                and visited[nrow][ncol]==False:
                    dfs(nrow,ncol)
        
        for r in range(ROWS):
            if visited[r][0]==False and board[r][0]=="O":
                    dfs(r,0)
            if visited[r][COLS-1]==False and board[r][COLS-1]=="O":
                    dfs(r,COLS-1)           

        for c in range(COLS):
            if visited[0][c]==False and board[0][c]=="O":
                    dfs(0,c)
            if visited[ROWS-1][c]==False and board[ROWS-1][c]=="O":
                    dfs(ROWS-1,c)           
        
        for r in range(ROWS):
            for c in range(COLS):
                if visited[r][c]==False and board[r][c]=="O":
                    board[r][c]="X"

# TC=O(m*n)
# SC=O(m*n)

File: Codes in C++/Graph/Surrounded Regions Replace 0s with Xs.cpp
class Solution {
public:
    void dfs(int row,int col,vector<vector<char>>& board,vector<vector<int>>& vis,int m,int n)
    {
        int delrow[]={-1,0,+1,0};
        int delcol[]={0,+1,0,-1};
        vis[row][col]=1;
        for(int i=0;i<4;i++)
        {
            int nrow=row+delrow[i];
            int ncol=col+delcol[i];
            if(nrow>=0 && nrow<m && ncol>=0 && ncol<n && 
            vis[nrow][ncol]==0 && board[nrow][ncol]=='O')
               dfs(nrow,ncol,board,vis,m,n); 
        }
    }

    void solve(vector<vector<char>>& board) {
       int m=board.size();
       int n=board[0].size(); 
       vector<vector<int>> vis(m,vector<int>(n,0));
       //Traverse first and last row
       for(int i=0;i<n;i++)
       {
           if(!vis[0][i]&&board[0][i]=='O')
                dfs(0,i,board,vis,m,n);
           if(!vis[m-1][i]&&board[m-1][i]=='O')
                dfs(m-1,i,board,vis,m,n);   
       } 
       
       //Traverse first and last column
       for(int i=0;i<m;i++)
       {
           if(!vis[i][0]&&board[i][0]=='O')
                dfs(i,0,board,vis,m,n);
           if(!vis[i][n-1]&&board[i][n-1]=='O')
                dfs(i,n-1,board,vis,m,n); 
       }

       for(int i=0;i<m;i++)
       {
           for(int j=0;j<n;j++)
           {
               if(!vis[i][j]&&board[i][j]=='O')
                    board[i][j]='X';
           }
       }

    }
};